File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    int count;
    public int goodNodes(TreeNode root) {
        
        if(root.left==null && root.right==null ){
            return 1;
        }
        
        count=0;
        int curr=root.val;
        
        helper(root,curr);
            
        return count;
    }
    
    public void helper(TreeNode root,int curr){
        
        if(root==null)
            return;
        
        if(root.val>=curr){
            count++;
            curr=root.val;
        }
        
        helper(root.left,curr);
        helper(root.right,curr);
    }
}

File: 1448_Count_Good_Nodes_in_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void good(TreeNode* root,int &count,int x){
        if(!root)return;
        if(root->val>=x){
            count++;
            x = root->val;
        }
        good(root->left,count,x);
        good(root->right,count,x);

    }
    
    int goodNodes(TreeNode* root) {
        int count =0;
        if(!root) return count;
        good(root,count,root->val);
        return count;
    }
};
